test/jdk/com/sun/jdi/TestScaffold.java

```diff
@@ -463,43 +463,65 @@ protected void failure(String str) {
+        // The example of args for line '@run driver Frames2Test -Xss4M' is  '-Xss4M' 'Frames2Targ'.
+        // The result without any wrapper enabled:
+        // argInfo.targetAppCommandLine : Frames2Targ
+        // argInfo.targetVMArgs : -Xss4M
+        // The result with wrapper enabled:
+        // argInfo.targetAppCommandLine : TestScaffold Virtual Frames2Targ
+        // argInfo.targetVMArgs : -Xss4M --enable-preview
+        boolean classNameParsed = false;
         for (int i = 0; i < args.length; i++) {
-            if (args[i].equals("-connect")) {
+            String arg = args[i].trim();
+            if (classNameParsed) {
+                // once classname is read, treat any other arguments as app arguments
+                argInfo.targetAppCommandLine += (arg + ' ');
+                continue;
+            }
+            if (arg.equals("-connect")) {
                 i++;
                 argInfo.connectorSpec = args[i];
-            } else if (args[i].equals("-trace")) {
+            } else if (arg.equals("-trace")) {
                 i++;
                 argInfo.traceFlags = Integer.decode(args[i]).intValue();
-            } else if (args[i].equals("-redefstart")) {
+            } else if (arg.equals("-redefstart")) {
                 redefineAtStart = true;
-            } else if (args[i].equals("-redefevent")) {
+            } else if (arg.equals("-redefevent")) {
                 redefineAtEvents = true;
-            } else if (args[i].equals("-redefasync")) {
+            } else if (arg.equals("-redefasync")) {
                 redefineAsynchronously = true;
-            } else if (args[i].startsWith("-J")) {
-                argInfo.targetVMArgs += (args[i].substring(2) + ' ');
+            } else if (arg.startsWith("-J")) {
+                argInfo.targetVMArgs += (arg.substring(2) + ' ');
 
                 /*
                  * classpath can span two arguments so we need to handle
                  * it specially.
                  */
-                if (args[i].equals("-J-classpath")) {
+                if (arg.equals("-J-classpath")) {
                     i++;
                     argInfo.targetVMArgs += (args[i] + ' ');
                 }
+            } else if (arg.startsWith("-")) {
+                argInfo.targetVMArgs += (arg + ' ');
             } else {
-                argInfo.targetAppCommandLine += (args[i] + ' ');
+                classNameParsed = true;
+                argInfo.targetAppCommandLine += (arg + ' ');
             }
         }
+
+        // Need to change args to run wrapper using command like 'TestScaffold Virtual <app-name>'
+        String mainWrapper = System.getProperty("main.wrapper");
+        if (mainWrapper != null && !argInfo.targetAppCommandLine.isEmpty()) {
+            argInfo.targetAppCommandLine = TestScaffold.class.getName() + ' '
+                    + mainWrapper + ' ' + argInfo.targetAppCommandLine;
+            argInfo.targetVMArgs += "--enable-preview ";
```

Comment:
It looks like previously we ignored `main.wrapper` if not set to `Virtual`, but with your chagnes we accept any setting. That's ok, but `--enable-preview` is really on needed when set to `Virtual`.